### ISSUE TITLE
Add checkstyle precommit for hidden/shadowed variables

### ISSUE THREAD

**Issue by dakrone** [elastic/elasticsearch#19752]
Checkstyle has the ability to detect when variables are shadowed by local variables: http://checkstyle.sourceforge.net/config_coding.html#HiddenField

We should enable this as it makes reading the code clearer, as well as reducing potential bugs with missing `this.` prefixes.

**Comment by jasontedor** [comment]
I assume that we would configure it so that this does not apply to constructor parameters nor to setter parameters?

**Comment by dakrone** [comment]
Yep, those are options we can definitely enable

**Comment by nik9000** [comment]
@dakrone, do you still think this is worth it?

**Comment by pugnascotia** [comment]
Do we still want to do this? I enabled the `HiddenField` rule and even after tweaking the settings, it finds a vast number of errors.

We could enable it as a warning, or an IDE-only warning?

**Comment by dakrone** [comment]
@pugnascotia I mean, *I* still want it, but I'm pretty opinionated about shadowing variables. Given that this is almost five years old, other people may have different opinions :D

**Comment by nik9000** [comment]
I guess there's two things:

* Does it find really bad stuff. yes, I know we're going to have differing opinions on what is bad. But. I think we'll mostly agree on *super bad*. Right?
* Is finding the super bad stuff worth the effort to remove the sorta-bad stuff. Or the pain of having another big list of checkstyle suppressions.

**Comment by dakrone** [comment]
> I think we'll mostly agree on *super bad*. Right?

For me shadowing is all about code readability. It's right up there with having a 140 character limit, sure we may not break things (well, in shadowing's case, we might), but it's so much easier to read code that doesn't shadow variables.

**Comment by mark-vieira** [comment]
@pugnascotia correct me if I'm wrong but I think we're all done here?

**Comment by pugnascotia** [comment]
Nope, I just ran out of steam with fixing up instances of shadowed vars. I can pick it up again though.
